dali/dfu/dfurun.cpp

```diff
@@ -1199,56 +1199,65 @@ class CDFUengine: public CInterface, implements IDFUengine
                                  foreignuserdesc->set(fu.str(),fp.str());
                             }
                             else
                                 foreignuserdesc.set(userdesc);
                         }
                     }
-                    if (foreigncopy) {
-                        foreignfdesc.setown(queryDistributedFileDirectory().getFileDescriptor(tmp.str(),foreignuserdesc,foreigndalinode));
-                        if (!foreignfdesc) {
-                            StringBuffer s;
-                            throw MakeStringException(-1,"Source file %s could not be found in Dali %s",tmp.str(),foreigndalinode->endpoint().getUrlStr(s).str());
-                        }
-                        kind.set(foreignfdesc->queryProperties().queryProp("@kind"));
-                        oldRoxiePrefix.set(foreignfdesc->queryProperties().queryProp("@roxiePrefix"));
-                        iskey = strsame("key", kind);
-                        if (destination->getWrap()||iskey)
-                            destination->setNumPartsOverride(foreignfdesc->numParts());
-                        if (options->getPush()) {// need to set ftslave location
+                    srcFile.setown(fdir.lookup(tmp.str(),userdesc,
+                            (cmd==DFUcmd_move)||(cmd==DFUcmd_rename)||((cmd==DFUcmd_copy)&&multiclusterinsert)));
+
+                    if (!srcFile)
+                        throw MakeStringException(-1,"Source file %s could not be found",tmp.str());
+
+                    iskey = isFileKey(srcFile);
+                    if ((cmd==DFUcmd_copy) && (srcFile->querySuperFile() != nullptr) && iskey)
+                        throwError1(DFTERR_InvalidSuperindexCopy, srcFile->queryLogicalName());
+
+                    oldRoxiePrefix.set(srcFile->queryAttributes().queryProp("@roxiePrefix"));
+                    kind.set(srcFile->queryAttributes().queryProp("@kind"));
+
+                    if (destination->getWrap()||iskey)
+                        destination->setNumPartsOverride(srcFile->numParts());
+
+                    if (options->getSubfileCopy())
+                        opttree->setPropBool("@compress",srcFile->isCompressed());
+
+                    if (foreigncopy)
+                    {
+                        foreignfdesc.setown(srcFile->getFileDescriptor());
+
+                        if (options->getPush())
+                        {
+                            // need to set ftslave location
                             StringBuffer progpath;
                             StringBuffer workdir;
-                            INode *n = foreignfdesc->queryNode(0);
-                            if (n&&getRemoteRunInfo("FTSlaveProcess", "ftslave", NULL, n->endpoint(), progpath, workdir, foreigndalinode, 1000*60*5)) {
+                            if (foreigndalinode && getRemoteRunInfo("FTSlaveProcess", "ftslave", NULL, foreigndalinode->endpoint(), progpath, workdir, foreigndalinode, 1000*60*5))
```

Comment:
this used to be:
```
                            INode *n = foreignfdesc->queryNode(0);
                            if (n&&getRemoteRunInfo("FTSlaveProcess", "ftslave", NULL, n->endpoint(), progpath, workdir, foreigndalinode, 1000*60*5)) {
 ```

i.e. it's using the endpoint() of the 1st node of the foreign file.
But now it's using the endpoint of the dalinode - which is not the same and incorrect.

Reply:
Changed